sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -103,9 +103,9 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
     // In case both sides have integral type, optimize the comparison by removing casts or
     // moving cast to the literal side.
     case be @ BinaryComparison(
-      Cast(fromExp, toType: IntegralType, _), Literal(value, literalType))
+      Cast(fromExp, toType: IntegralType, tz), Literal(value, literalType))
         if canImplicitlyCast(fromExp, toType, literalType) =>
-      simplifyIntegralComparison(be, fromExp, toType, value)
+      simplifyIntegralComparison(be, fromExp, toType, value, tz)
```

Comment:
This looks like irrelevant as a follow-up.
Is there a comment on this? `pass timezone info to the generated cast on the literal value`?

Reply:
It's hard for me to think a valid test case for this in the scope of SPARK-24994 (Integral Types). Although Spark handles some time-related types with the integral types, I believe we need to revisit this on new PR which aims to support time.

Reply:
Thanks @dongjoon-hyun . I discovered this in tests - without this some of the existing tests will fail because of timezone mismatch (although I think this shouldn't affect correctness given we only handle integral types?).

In the follow-up, the current plan is to only extend to numeric types (e.g., float/double/decimal). I don't know whether support for time-related types is important or not.

Reply:
Could you give me some pointer for the failures? Or, can we have a reproducer? Usually, every code change had better have a test coverage. If there is a test case for this, it sounds reasonable.

Reply:
（oops just found out my comment was not sent out successfully)

This is because `ResolveTimeZone` will try to add timezone info to all expressions that don't have it during query analysis. However, since the `Cast` expr was generated at optimization phase, it will not have the timezone info. As result, `PlanTest.comparePlans` will fail because of mismatch. I can try to come up with a test if necessary.

On the other hand, I think instead of using `Cast`, we may just directly use the value, since the `Cast` will be optimized away by ConstantFolding anyways later. What do you think?

Reply:
yea this also works: just evaluate the cast and make a literal.

Reply:
Thanks. I'll do that then. I'm not totally sure about`Cast.canonicalize` - somehow it is not used in query analysis and optimization and the lazy val is not initialized before the plan comparison.